wip-trampoline.md

```diff
@@ -0,0 +1,840 @@
+`option_trampoline_routing`. This is an opportunity for them to earn more fees
+than with the default onion routing.
+
+A payer selects a few trampoline nodes and builds a corresponding trampoline
+onion TLV. It then embeds that trampoline onion in the last `per_hop` payload
+of an `onion_packet` destined to the first trampoline node. Computing routes
+between trampoline nodes is deferred to the trampoline nodes themselves.
+
+Merchants may include a few trampoline nodes that are close to them in their
+invoice to help payers select a good last trampoline node. This is optional and
+payers aren't required to use this routing hint.
+
+An interesting side-effect is that trampoline routing combines nicely with
+rendezvous routing. With a simple new TLV type and more data in the invoice,
+rendezvous routing can easily work on top of trampoline routing, with a
+construction similar to what is described [here][rendezvous].
+
+## Table of Contents
+
+* [Packet Structure](#packet-structure)
+  * [Trampoline Onion](#trampoline-onion)
+  * [Usage](#usage)
+  * [Paying to non-trampoline nodes](#paying-to-non-trampoline-nodes)
+* [Fees and CLTV requirements](#fees-and-cltv-requirements)
+* [Network Pruning](#network-pruning)
+  * [The `node_update` Message](#the-node_update-message)
+  * [Fees and CLTV estimation](#fees-and-cltv-estimation)
+  * [Filtering gossip messages](#filtering-gossip-messages)
+* [Examples](#examples)
+  * [Merchant supporting trampoline payments](#merchant-supporting-trampoline-payments)
+  * [Merchant without trampoline support](#merchant-without-trampoline-support)
+* [References](#references)
+
+## Packet Structure
+
+### Trampoline Onion
+
+The trampoline onion is a fixed-size TLV packet with the following structure:
+
+1. type: `trampoline_onion_packet`
+2. data:
+   * [`1`:`0x08`] (`type`)
+   * [`3`:`0xfde202`] (`length`)
+   * [`1`:`version`]
+   * [`33`:`public_key`]
+   * [`672`:`trampoline_hops_data`]
+   * [`32`:`hmac`]
+
+This results in a `742`-bytes fixed-size packet where the value has exactly the
+same format as the `onion_packet` with a smaller `hops_data`.
+
+The `trampoline_hops_data` field is a list of `trampoline_hop_payload`s. A
+`trampoline_hop_payload` is a structure that holds obfuscations of the next
+trampoline hop's address, transfer information, and its associated HMAC. It
+uses the same format as the `onion_packet`'s variable-size `per_hop` payload,
+meaning that it can contain an arbitrary TLV stream:
+
+1. type: `trampoline_hops_data`
+2. data:
+   * [`n1`:`trampoline_hop_payload`]
+   * [`n2`:`trampoline_hop_payload`]
+   * ...
+   * `filler`
+
+Each `trampoline_hop_payload` has the following structure:
+
+1. type: `trampoline_hop_payload`
+2. data:
+   * [`var_int`:`length`]
+   * [`length`:`tlv_stream`]
+   * [`32`:`hmac`]
+
+The `tlv_stream` contains data necessary to route to the next trampoline hop.
+The following types MAY be used to encode that data:
```

Comment:
It feels like many TLV types will be useful accross many message types.
I'd like to have a centralized place where we could register known TLVs and their assigned `type` (like what we have in bolt #9). I think that all the TLVs I'm using here (except the `trampoline_onion`) will be useful to other messages and should be specified outside of this PR.